source/Locales/ScottLocale/ScottLocalesOfAlgebraicDcpos.lagda

```diff
@@ -0,0 +1,361 @@
+Ayberk Tosun.
+
+Started on 29 September 2023
+Finished on 2 October 2023.
+
+This module contains the definition of the Scott locale of a large and locally
+small _algebraic_ dcpo, using the definition of algebraic dcpo from the
+`DomainTheory` development due to Tom de Jong.
+
+If one starts with an algebraic dcpo, one can ensure that the resulting Scott
+locale is small by quantifying over only the basic/compact opens. This is the
+difference between the construction in this module and the one in
+`ScottLocale.Definition`
+
+TODO: in the future, it would be good to call the other module something
+else other than `Definition`.
+
+\begin{code}[hide]
+
+{-# OPTIONS --safe --without-K --exact-split #-}
+
+open import MLTT.Spartan
+open import Slice.Family
+open import UF.FunExt
+open import UF.Logic
+open import UF.PropTrunc
+open import UF.SubtypeClassifier
+open import UF.Subsingletons
+open import UF.Subsingletons-FunExt
+open import UF.Powerset-MultiUniverse hiding (_⊆_)
+
+\end{code}
+
+We assume the existence of propositional truncations as well as function
+extensionality, and we assume a universe 𝓤 over which we consider large and
+locally small dcpos.
+
+\begin{code}
+
+module Locales.ScottLocale.ScottLocalesOfAlgebraicDcpos
+        (pt : propositional-truncations-exist)
+        (fe : Fun-Ext)
+        (𝓤  : Universe)
+        where
+
+open Universal fe
+open Implication fe
+open Existential pt
+open Conjunction
+
+open import Locales.Frame pt fe
+
+open import DomainTheory.Basics.Dcpo pt fe 𝓤 renaming (⟨_⟩ to ⟨_⟩∙)
+open import DomainTheory.Topology.ScottTopology pt fe 𝓤
+open import DomainTheory.BasesAndContinuity.Continuity pt fe 𝓤
+open import DomainTheory.BasesAndContinuity.Bases pt fe 𝓤
+
+open import Locales.ScottLocale.Definition pt fe 𝓤
+
+open PropositionalTruncation pt
+
+\end{code}
+
+The construction is carried out over a large and locally small dcpo 𝓓 equipped
+with a small compact basis i.e. is _structurally algebraic_ in the terminology
+of Tom de Jong. I also use the term "structure of algebraicity" for this.
+
+\begin{code}
+
+module ScottLocaleConstruction (𝓓    : DCPO {𝓤 ⁺} {𝓤})
+                               (hscb : has-specified-small-compact-basis 𝓓)
+                               (pe   : propext 𝓤)                          where
+
+ open import DomainTheory.Lifting.LiftingSet pt fe 𝓤 pe
+ open DefnOfScottTopology 𝓓 𝓤
+ open DefnOfScottLocale 𝓓 𝓤 pe using (𝒪ₛ-equality)
+
+\end{code}
+
+We denote by `𝕒` the fact that the dcpo `𝓓` in consideration is _structurally
+algebraic_.
+
+\begin{code}
+
+ 𝕒 : structurally-algebraic 𝓓
+ 𝕒 = structurally-algebraic-if-specified-small-compact-basis 𝓓 hscb
+
+\end{code}
+
+We denote by `I` the index type of the basis and by `β` its enumeration
+function.
+
+\begin{code}
+
+ I = pr₁ hscb
```

Comment:
Could we name the projections for clarity and modularity and portability?

Reply:
The reason I used a projection here was actually to _avoid projections_ in the module. I couldn't find a standard projection for this, and I didn't want to modify the `DomainTheory` modules. I'm not sure what the best thing to do here is and I always think about this.

I will look at `DomainTheory.BasesAndContinuity.Bases` more carefully to find out if there's a record-based definition that I can use for this purpose, but that might not be the case. In situations like these, are you fine with me adding record-based definitions for things (cc @tomdjong)?

Reply:
It's fine with me to add some records.

Reply:
>  It's fine with me to add some records.

I will do this in a future PR @martinescardo. I don’t want to add too many tangential changes into this PR. I 100% agree that I should completely avoid projections. If there were a well-principled way for me to avoid projections (i.e. enforcing this using the type system), I would use it.

In the presence of an ML-like module system, for example, I would keep the `Σ`-based definition abstract, so that projections are literally not available to me, and I would expose only the equivalence with the record-based definition, so that this is the only way to make use of an inhabitant of the `Σ`-based definition. In Agda, one could use typeclasses to do this perhaps, but I’m not sure if it’s a good idea.